Arch review (#1863)
giantswarm/docs: src/content/platform-overview/architecture/index.md

---

> linkTitle: Platform architecture

Author: I have simplified this document to be readable. Before it has a long list of components and various information. My goal is to explain the Platform concept and how is the Architecture without going to much into details

owner:
  - https://github.com/orgs/giantswarm/teams/team-rocket
  - https://github.com/orgs/giantswarm/teams/team-phoenix
  - https://github.com/orgs/giantswarm/teams/team-horizon
---

Giant Swarm's mission is to offer a developer platform to our customers. We work with companies which usually have their systems and teams managing their existing infrastructure and applications. For that reason, we have created a composable system where customers can opt in or out of the services available. 

All platforms teams are trying to solve the same set of problems. How developers securely build their applications, how to deploy these applications in an ephemeral environment easily or how to observe the behavior of these applications with the minimal configuration. For that reason we want to build a platform that relies on the [good principles](https://www.giantswarm.io/blog/platform-engineering-its-not-about-a-tool-stack-its-a-set-of-capabilities) exposed by the community:

- Treat the platform as product
- Minimize cognitive load for developers
- Enable fast-flow software delivery

![Platform architecture](platform-architecture.png)

> Starting from this premise we have built a solution that consists in various systems. They can be categorized into three areas: infrastructure, applications, platform interfaces and operations.

Author: no worries this is marvellous, thanks for such addition!

In addition to the security policies, [Network Policies](https://kubernetes.io/docs/concepts/services-networking/network-policies/) define the communication policies to and from the applications in each namespace. All components to run a cluster provided by Giant Swarm come with strict policies by default. Our managed namespaces (“kube-system” and “giantswarm”) block all traffic in general, so only expected and specifically configured routes and ports are enabled. Customers can follow this approach and deny all communications by default in their application namespaces forcing each workload to define which communications are allowed. This [guide]({{< relref "/getting-started/network-policies" >}}) helps to understand how such a dynamic firewall works.

Currently we run [Cilium](https://cilium.io/) as CNI in all our clusters which brings new powerful features to increase the security of our setups. Cilium extends Kubernetes network policies to allow us filter out based on domains and

*Monitoring Features of the Platform*

> Since we provide a **managed** Kubernetes platform, Giant Swarm has to be aware of state and unexpected events regarding the platform. For that reason our management clusters run a [monitoring stack](https://www.giantswarm.io/blog/monitoring-on-demand-kubernetes-clusters-with-prometheus) to watch all workload clusters and ensure all managed components are healthy. In each workload cluster there are several [exporters](https://prometheus.io/docs/instrumenting/exporters/) that gather and forward the metrics for each component.

Reviewer: ```suggestion
Since we provide a **managed** Kubernetes platform, Giant Swarm keeps a close eye on the state of the platform, and is alerted for various types of unexpected events. To support our 24/7 operations, our management clusters run a [monitoring stack](https://www.giantswarm.io/blog/monitoring-on-demand-kubernetes-clusters-with-prometheus) to watch all workload clusters and ensure all managed components are healthy. In each workload cluster, there are several [exporters](https://prometheus.io/docs/instrumenting/exporters/) that gather and forward the metrics for each component.
```

> Our on-call engineers are paged in case anything happens to the cluster or its base components. They respond to the incident based on our run-books, the ones we have written (and continue to update) over years operating Cloud Native systems. In case there is an improvement to be made, a post mortem is created and a solution will be implemented before long. Any patch or fix added to the platform is released to all customers.

Reviewer: ```suggestion
Our on-call engineers are paged in case anything happens to the cluster or its base components. They respond to the incident based on our run-books, which we have written (and continue to update) over years operating Cloud Native systems. We perform post-mortem analyses to identify the cause of platform issues or outages, and to identify opportunities to improve the platform's resiliency. In case there is an improvement to be made, a solution will be implemented by the same team which is responsible for operating it. Any patch or fix added to the platform is then released to all customers.
```

*CI/CD Features of the Platform*

> Since the appearance of [GitOps](https://www.giantswarm.io/blog/what-is-gitops) we have been enthusiastic about it. It provides many benefits while relying on the same principles we were already advocating for. In Giant Swarm, we use [Flux](https://www.giantswarm.io/blog/gitops-with-flux-giant-swarm) to control the configuration and definition of infrastructure and the software on top of it.

Reviewer: ```suggestion
Since the appearance of [GitOps](https://www.giantswarm.io/blog/what-is-gitops) we have been enthusiastic about it. It provides many benefits while relying on the same principles we were already advocating for. Giant Swarm uses [Flux](https://www.giantswarm.io/blog/gitops-with-flux-giant-swarm) to control the configuration and definition of infrastructure and the software on top of it.
```

> In our setup we have two Flux instances running, one managing the resources specific to the overall operation of the management cluster (`flux-giantswarm`) and the second for handling resources generated by the customer (`flux-system`).

Reviewer: ```suggestion
In our setup, we have two Flux instances running, one managing the resources specific to the overall operation of the management cluster (`flux-giantswarm`) and the second for handling resources generated by the customer (`flux-system`).
```

There are several entrypoints to the platform via our largely Kubernetes-based interfaces.

We leverage Kubernetes and its extension capabilities to expose the main functionality. The [management API]({{< relref "/platform-overview/management-api" >}}) is just a Kubernetes API enriched to enable to serve complete golden paths for developers.

Beyond the management API, we recommend a [GitOps](https://www.giantswarm.io/blog/what-is-gitops)-based approach to ensure customers use solid principles for managing their workloads. Virtually all features of our platform can be described, configured, and stored in your repositories, maintaining a declarative source of truth.

Along with GitOps we have a [Web interface]({{< relref "/platform-overview/web-interface" >}}), that simplifies the life of the platform engineering teams in order to visualize infrastructure, apps, and permissions across the entire platform. 

In addition, we offer a set of [templates](https://github.com/giantswarm/gitops-template) to start with the platform and a [documentation hub]({{< relref "/getting-started" >}}) including a step-by-step guide to completing a Dev Platform Journey. 

### Applications

> The capabilities of the platform are supported thanks to the fantastic selection Cloud Native tools available. At Giant Swarm, we use and manage a set of open-source tooling backed by the community, but we do not force you to stick to it. 

Reviewer: Opinion: I know some folks love that "best of breed" but it kind of bothers me :see_no_evil:

There is a set of capabilities that customers can rely on today, but we are also continuously on the hunt for new use cases. Every customer has different needs, which evolve over time, and we intend for the platform to evolve in kind.

Today we can offer a set of different features for building your platform, including:

- Access Management: Configure which users or groups can have access to the platform.
- Secret Management: Securely store the secrets your apps need to access other services.
- CI/CD: Create your pipelines to build and deploy the applications.
- Registry: Store application artifacts in a secure and available place.
- Smart Routing: Configure ingress/egress access to your applications with advanced service mesh capabilities.
- Policy Enforcement: Ensure compliance with company policies across the platform landscape.
- Resource provisioning: Enable easy provisioning and configuration of external resources.

### Infrastructure

> For managing the infrastructure, we run a management cluster per provider and region where you want to have your workloads. From that management cluster, you can spin up as many individual Kubernetes clusters, called workload clusters, as you want. Our operations team works to maintain all cluster components' health, and continuously releases new versions with new features and patches.

Reviewer: Should be "at will"

> For managing the infrastructure, we run a management cluster per provider and region where you want to have your workloads. From that management cluster, you can spin up as many individual Kubernetes clusters, called workload clusters, as you want. Our operations team works to maintain all cluster components' health, and continuously releases new versions with new features and patches.

Reviewer 2: fixed

![Cluster architecture image](CAPI_architecture.png)

By default, the machines are spread across three different failure domains, or zones, to ensure the availability of the API and workloads running on top. In our setup, three control plane machines hold the Kubernetes API and the other controllers, and a variable number of worker machines contain the workloads.

Besides the Kubernetes machines, we run a bastion host that helps us with the operations. It is the single entry point to the running infrastructure. That way all cluster machines can live in a private network and reduce the exposure of the services running on them via explicit configuration. We are working on a [Teleport](https://goteleport.com/) solution to replace the necessity of the bastions and centralize cluster access through a single tool.

#### Providers

Although we try to maintain feature parity between infrastructure providers, offering the same features and using the same configuration, there are some subtle differences depending on the underlying provider functionality.

{{< tabs >}}
{{< tab id="flags-clouddirector" title="VMware Cloud Director">}}

> *Compatibility*

Reviewer 3: Let's use h5 headings here

Reviewer 3: and maybe CSS rule `.tab-pane > h5:first-child { margin-top: 0 }` so it doesn't look awkward

> The setup supports organization virtual datacenters (OVCDs) running on VMware Cloud Director 10.3 and above. It must be backed by backed with NSX-T and NSX advanced load balancer (ALB) with the load balancer feature enabled on the Edge gateway. 

Reviewer 2: ```suggestion
The setup supports organization virtual datacenters (OVCDs) running on VMware Cloud Director 10.3 and above. It must be backed by NSX-T and NSX advanced load balancer (ALB) with the load balancer feature enabled on the Edge gateway. 
```

*Authentication*

> Cluster API Provider VMware Cloud Director (CAPVCD), along with the associated Cloud Provider interface (CPI) and Container Storage interface (CSI), authenticate against the VMware Cloud Director API using an API Token (sometimes also referred to as Refresh Token) which is stored in a secret. Such a token can be created by any user with the right permission and can be revoked at any time should there be suspicion of it being compromised.

Reviewer 2: ```suggestion
Cluster API Provider VMware Cloud Director (CAPVCD), along with the associated Cloud Provider interface (CPI) and Container Storage interface (CSI), authenticate against the VMware Cloud Director API using an API Token (sometimes also referred to as Refresh Token) which is stored in a secret. Such token can be created by any user with the right permissions and can be revoked at any time, should there be suspicion of it being compromised.
```

*Networking*

> The kubernetes API and services of type `LoadBalancer` get IPs from a pool of external IPs available in the edge gateway. It can be set statically or takes the next available IP if unspecified. A virtual service is then created with the required IP/port and is associated with a load balancer pool that contains the relevant node IPs as members. For the CPI, we support the virtual service shared feature which was introduced in VCD 10.4 as well as the legacy method based on a single internal IP and multiple DNAT rules.

Reviewer 2: ```suggestion
The Kubernetes API and services of type `LoadBalancer` get IPs from a pool of external IPs available in the edge gateway. It can be set statically or takes the next available IP if unspecified. A virtual service is then created with the required IP/port and is associated with a load balancer pool that contains the relevant node IPs as members. For the CPI, we support the virtual service shared feature which was introduced in VCD 10.4 as well as the legacy method based on a single internal IP and multiple DNAT rules.
```

{{< /tab >}}
{{< tab id="flags-aws" title="AWS">}}

The setup uses a standalone VPC (though you can bring your own VPC) and creates private subnets for the machine in each availability zone. It uses NAT gateways for allowing machines to pull images or route requests from containers to the Internet. On the other side, it needs an Internet Gateway to route traffic from Internet to the containers. It leverages route tables to configure the routing for each subnet and gateways.

{{< /tab >}}
{{< tab id="flags-capz" title="Azure">}}

{{< /tab >}}
{{< /tabs >}}

Internally, Cluster API (CAPI) uses kubeadm to configure all the machines according to current standards. It uses a template defined as yaml, part of CAPI, where we have hardened the different parameters for the API and other controllers running in the master machine.

### Operations

Aside from infrastructure and customer-facing applications, we also deploy a set of tooling that ensure we have a seamless delivery system, proper hardening, and clear observability.  

> *Authentication Features of the Platform*

Reviewer 3: Let's use h4 headings here

Giant Swarm configures clusters to be secure by default. [Role-Based Access Control (RBAC)](https://kubernetes.io/docs/reference/access-authn-authz/rbac/) is enabled by default and our customers can create their own roles or use the ones predefined in the cluster to gain access to manage their workloads. The concept of authenticated users and groups does not exist in Kubernetes, so it relies on an external solution to retrieve user/group information (e.g. via X.509 certificates or [OIDC](https://en.wikipedia.org/wiki/OpenID_Connect)). Although our platform allows users to access their clusters using certificates, we recommend using an OIDC compliant Identity Provider, such as Active Directory, to provide authentication. There are several advantages to using an OIDC provider, such as short lived tokens and taking advantage of existing user and group information. Once authentication is sorted out, the authorization part is handled with RBAC. RBAC, along with namespaces, lets users define granular permissions for each user or group. This [guide]({{< relref "/getting-started/rbac-and-psp" >}}) will walk you through it.

*Secure Features of the Platform*

> Since our earliest releases, Giant Swarm has set up a secure baseline in all our customer clusters. In the early days, Kubernetes supported [Pod Security Policies (PSPs)](https://kubernetes.io/docs/concepts/policy/pod-security-policy/) to enforce pod security, which provided a built-in resource for a user to define permitted pod configurations or volume types. In Kubernetes `1.25`, this implementation has been phased out in favor of [Pod Security Admission(PSA)](https://kubernetes.io/docs/concepts/security/pod-security-admission/). We have not been able to maintain an equivalent set of capabilities using that technology, so, for now, we have [decided to leverage on Kyverno to enforce appropriately restricted and extensible policies](https://www.giantswarm.io/blog/giant-swarms-farewell-to-psp). By default, users and workloads running in Giant Swarm clusters are bound by a restrictive policy that disallows, among many others, behaviors like running containers as root or mounting host path volumes. Cluster operators must grant applications higher security privileges on a case by case basis.

Reviewer 2: ```suggestion
Since our earliest releases, Giant Swarm has set up a secure baseline in all our customer clusters. In the early days, Kubernetes supported [Pod Security Policies (PSPs)](https://kubernetes.io/docs/concepts/policy/pod-security-policy/) to enforce pod security, which provided a built-in resource for a user to define permitted pod configurations or volume types. In Kubernetes `1.25`, this implementation has been phased out in favor of [Pod Security Admission(PSA)](https://kubernetes.io/docs/concepts/security/pod-security-admission/). We have not been able to maintain an equivalent set of capabilities using that technology, so, for now, we have [decided to leverage Kyverno to enforce appropriately restricted and extensible policies](https://www.giantswarm.io/blog/giant-swarms-farewell-to-psp). By default, users and workloads running in Giant Swarm clusters are bound by a restrictive policy that disallows, among many others, behaviors like running containers as root or mounting host path volumes. Cluster operators must grant applications higher security privileges on a case by case basis.
```
